crates/ruma-events/src/call/matrix_rtc.rs

```diff
@@ -0,0 +1,409 @@
+//! Types for matrixRTC state events ([MSC3401]).
+//!
+//! This implements a newer/updated version of MSC3401.
+//!
+//! [MSC3401]: https://github.com/matrix-org/matrix-spec-proposals/pull/3401
+
+#[cfg(feature = "unstable-msc4075")]
+pub mod notify;
+
+pub mod focus;
+pub mod member_data;
+pub mod member_event;
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use ruma_common::MilliSecondsSinceUnixEpoch as TS;
+    use ruma_events::{
+        call::matrix_rtc::notify::{ApplicationType, CallNotifyEventContent, NotifyType},
+        Mentions,
+    };
+    use serde_json::{from_value as from_json_value, json, to_value as to_json_value};
+
+    use super::{
+        focus::{ActiveFocus, ActiveLivekitFocus, Focus, LivekitFocus},
+        member_data::{
+            Application, CallApplicationContent, CallScope, LegacyMembershipData, MembershipData,
+        },
+        member_event::MemberEventContent,
+    };
+
+    fn create_call_member_legacy_event_content() -> MemberEventContent {
+        MemberEventContent::new_legacy(vec![LegacyMembershipData {
+            application: Application::Call(CallApplicationContent {
+                call_id: "123456".to_owned(),
+                scope: CallScope::Room,
+            }),
+            device_id: "ABCDE".to_owned(),
+            expires: Duration::from_secs(3600),
+            foci_active: vec![Focus::Livekit(LivekitFocus {
+                alias: "1".to_owned(),
+                service_url: "https://livekit.com".to_owned(),
+            })],
+            membership_id: "0".to_owned(),
+            created_ts: None,
+        }])
+    }
+
+    fn create_call_member_event_content() -> MemberEventContent {
+        MemberEventContent::new(
+            Application::Call(CallApplicationContent {
+                call_id: "123456".to_owned(),
+                scope: CallScope::Room,
+            }),
+            "ABCDE".to_owned(),
+            ActiveFocus::Livekit(ActiveLivekitFocus {
+                focus_select: super::focus::FocusSelection::OldestMembership,
+            }),
+            vec![Focus::Livekit(LivekitFocus {
+                alias: "1".to_owned(),
+                service_url: "https://livekit.com".to_owned(),
+            })],
+            None,
+        )
+    }
+
+    #[test]
+    fn serialize_call_member_event_content() {
```

Comment:
Could we have also tests for serializing/deserializing a whole event? Only for the non-legacy version would be enough.

Reply:
De-serializing makes a lot of sense. I cannot find the Serializable version of a full event however.
AnyStateEvent and OriginStateEvent are all not serializable.

Reply:
Yeah sorry, I forgot for a minute that it's not possible to serialize full events.

Reply:
I did the de-serialization for the full event. Which makes a lot of sense with the custom Redacts types.